### PR TITLE
[WIP] fix: fixed ConnPool incorrect handling of transaction

### PR DESCRIPTION
**Pull request by jiekun**
- [X] Do only one thing
- [X] Non breaking API changes
- [X] Tested

### What did this pull request do?
Currently, sharding lib cannot start transaction correctly.

1. When entering  `(pool *ConnPool) BeginTx()` method, `pool.ConnPool` is actually `*sql.DB` and not match the `gorm.ConnPoolBeginner` interface. So `pool, nil` is returned and transaction is not started.
2. `BeginTx` / `Commit` / `Rollback` return nil when interface not match. When executing write operation, transaction is automatically called: [`BeginTransaction`](https://github.com/go-gorm/gorm/blob/v1.23.5/callbacks/transaction.go#L7). If no error returned, it will consider the transaction started by GORM and try to commit it in `CommitOrRollbackTransaction`.

In this PR, problems above are fixed by:
1. Adding correct interface assertion for `BeginTx` so the `BeginTx` could be called.
2. return `gorm. ErrInvalidTransaction` instead of nil, which could be handled in [`transaction.go`](https://github.com/go-gorm/gorm/blob/v1.23.5/callbacks/transaction.go#L12)

### User Case Description
N/A

### CODE REVIEW

**Comment by jiekun** [comment]
@jinzhu @huacnlee PTAL

**Comment by jiekun** [comment]
#43 #41 Should be fixed in this PR. We may need different struct to impl `TxCommitter`

**Review by huacnlee — CHANGES_REQUESTED** [review]

File: conn_pool.go [hunk 1]
```diff
@@ -61,40 +66,105 @@ func (pool ConnPool) QueryContext(ctx context.Context, query string, args ...int
...
 	}
 
-	return pool, nil
+	return pool, gorm.ErrInvalidTransaction
 }
 
-// Implement TxCommitter.Commit
-func (pool *ConnPool) Commit() error {
+// Commit Implement TxCommitter.Commit
+func (pool *ShardTxCommitter) Commit() error {
```
[the remark this answers is not in the document]
Comment (jiekun) [reply conn_pool.go:93]: But ConnPool struct should NOT implement Commit and Rollback method. Otherwise the nested transaction wont work.

**Comment by jiekun** [comment]
Hi. I'm closing this PR due to the lack of time to review and push it further. The issue is still present as described above.

**Pull request closed without merging**